govuk_chat_evaluation/rag_answers/evaluate.py

```diff
@@ -1,31 +1,227 @@
+import logging
+import os
 from pathlib import Path
+from typing import cast, Any
+from dataclasses import asdict
+from functools import cached_property
+from pydantic import BaseModel
+import pandas as pd
 
 from deepeval import evaluate as deepeval_evaluate
-from deepeval.metrics import GEval
-from deepeval.test_case import LLMTestCaseParams
+from deepeval.evaluate import TestResult
+from deepeval.test_case import LLMTestCase, MLLMTestCase
+from deepeval.metrics import BaseMetric
 
+
+from ..timing import Timer
 from ..file_system import jsonl_to_models
-from .data_models import EvaluationTestCase
+from .data_models import EvaluationTestCase, EvaluationConfig
+
+
+logger = logging.getLogger(__name__)
 
 
 # would expect we need to pass config object through if that has metrics configuration
-def evaluate_and_output_results(_output_dir: Path, evaluation_data_path: Path):
-    models = jsonl_to_models(evaluation_data_path, EvaluationTestCase)
-    run_deepeval_evaluation(models)
+def evaluate_and_output_results(
+    _output_dir: Path, 
+    evaluation_data_path: Path, 
+    evaluation_config: EvaluationConfig, 
+    deepeval_evaluate_params: dict
+):
+    """
+    Function to run the evaluation, aggregate the results, and export them to files.
 
+    Args:
+        _output_dir: The directory to save the evaluation results.
+        evaluation_data_path: Path to the JSONL file containing the evaluation data.
+        evaluation_config: Configuration for the evaluation.
+        deepeval_evaluate_params: Parameters to be passed to deepeval evaluation.
+    """
+    # set DeepEval results folder
+    os.environ["DEEPEVAL_RESULTS_FOLDER"] = str(_output_dir)
 
-def run_deepeval_evaluation(models: list[EvaluationTestCase]):
-    llm_test_cases = [model.to_llm_test_case() for model in models]
+    # convert raw data into model instances for evaluation
+    models = jsonl_to_models(evaluation_data_path, EvaluationTestCase)
 
-    # contrived example metric
-    correctness_metric = GEval(
-        name="Correctness",
-        criteria="Determine if the 'actual output' is correct based on the 'expected output'.",
-        evaluation_params=[
-            LLMTestCaseParams.ACTUAL_OUTPUT,
-            LLMTestCaseParams.EXPECTED_OUTPUT,
-        ],
-        threshold=0.5,
+    # initialise EvaluationResults with the config and the cases
+    evaluation_results = EvaluationResults(
```

Comment:
I don't think this class name is quite right to reflect the purpose - as rather than being a representation of EvaluationResults it's more a mechanism to generate the results - so something like an EvaluationRunner perhaps?

I'd wonder if you did actually want this to be a class though and might want to create a module deepeval_evaluate.py that had an imported function into here. Producing code here of something like:
```
models = jsonl_to_models(evaluation_data_path, EvaluationTestCase)
evaluation_results = run_deepeval_evaluation(config=evaluation_config, cases=[model.to_llm_test_case() for model in models])
aggregation = AggregatedResults(evaluation_results)
```

Reply:
Yes, I agree. I had a hard time trying to fit the orchestration into an `EvaluationResults` and `AggregationResults` class, following what is in the jailbreak guardrail, as I (mis)understood that was the general framework to respect.

I originally developed `run_deepeval_evaluation` to be a standalone function (https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/18/commits/3fd49752a83497fc9d53efba515372c10e0d7a71).

Let's chat offline how to redesign this!

Reply:
Ah sure, so for JailbreakGuardrails we have EvaluationResult singular rather than EvaluationResults, so that represents an individual check, whereas AggregateResults are the collective. So you have a one to many of one AggregateResults summarising multiple EvaluationResult objects. 

Yup let's find a time to chat - I imagine functions alone will suffice as I think it's mostly a procedure rather than something with state.